src/bin/pg_autoctl/supervisor.c

```diff
@@ -0,0 +1,651 @@
+ * file.
+ */
+static bool
+supervisor_loop(pid_t start_pid,
+				Service services[],
+				int serviceCount,
+				const char *pidfile)
+{
+	int subprocessCount = serviceCount;
+	int stoppingLoopCounter = 0;
+	bool shutdownSequenceInProgress = false;
+	bool cleanExit = false;
+
+	/* wait until all subprocesses are done */
+	while (subprocessCount > 0)
+	{
+		pid_t pid;
+		int status;
+
+		/* Check that we still own our PID file, or quit now */
+		(void) check_pidfile(pidfile, start_pid);
+
+		/* If necessary, now is a good time to reload services */
+		if (asked_to_reload)
+		{
+			int serviceIndex = 0;
+
+			for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+			{
+				Service *service = &(services[serviceIndex]);
+
+				log_debug("Reloading pg_autoctl %s service", service->name);
+
+				(void) (*service->reloadFunction)(service);
+			}
+
+			/* reset our signal handling facility */
+			asked_to_reload = 0;
+		}
+
+		/* ignore errors */
+		pid = waitpid(-1, &status, WNOHANG);
+
+		switch (pid)
+		{
+			case -1:
+			{
+				if (errno == ECHILD)
+				{
+					/* no more childrens */
+					if (asked_to_stop || asked_to_stop_fast)
+					{
+						/* off we go */
+						log_info("Internal subprocesses are done, stopping");
+						return true;
+					}
+				}
+				else
+				{
+					log_fatal("Failed to call waitpid(): %m");
+					return false;
+				}
+			}
+
+			case 0:
+			{
+				/*
+				 * We're using WNOHANG, 0 means there are no stopped or exited
+				 * children, it's all good. It's the expected case when
+				 * everything is running smoothly, so enjoy and sleep for
+				 * awhile.
+				 */
+				if (asked_to_stop || asked_to_stop_fast)
+				{
+					cleanExit = true;
+
+					/*
+					 * Stop all the services.
+					 */
+					if (stoppingLoopCounter == 0)
+					{
+						(void) supervisor_stop_subprocesses(services,
+															subprocessCount);
+					}
+
+					++stoppingLoopCounter;
+
+					if (stoppingLoopCounter == 1)
+					{
+						log_info("Waiting for subprocesses to terminate.");
+					}
+
+					/*
+					 * If we've been waiting for quite a while for
+					 * sub-processes to terminate. Let's signal again all our
+					 * process group ourselves and see what happens next.
+					 */
+					else if (stoppingLoopCounter == 50)
+					{
+						log_info("pg_autoctl services are still running, "
+								 "signaling them with SIGTERM.");
+
+						if (!supervisor_signal_process_group(SIGTERM))
+						{
+							log_warn(
+								"Still waiting for subprocesses to terminate.");
+						}
+					}
+
+					/*
+					 * Wow it's been a very long time now...
+					 */
+					else if (stoppingLoopCounter > 0 &&
+							 stoppingLoopCounter % 100 == 0)
+					{
+						log_info("pg_autoctl services are still running, "
+								 "signaling them with SIGINT.");
+
+						if (!supervisor_signal_process_group(SIGINT))
+						{
+							log_warn(
+								"Still waiting for subprocesses to terminate.");
+						}
+					}
```

Comment:
Good that you updated the comment, but I still think this code is pretty strange in general. The main problems I have in general:
1. If the user asked for a normal shutdown, at some point we will automatically upgrade it to a fast shutdown.
2. We assume these signals work as we expect on the child processes. Which is not necessarily the case, SIGINT, SIGTERM and SIGQUIT have different meaninings on postgres and us.
3. These timeouts cannot be configured.

This is why I think it's better to simply require the user to send a different signal after a different timeout, if the process has not quit yet. I believe this is configurable in both systemd and docker. This does require us to handle an update to the signal, i.e. if the user first sends SIGTERM and then SIGQUIT we still handle the SIGQUIT. Which I think is good to do anyway.

Reply:
I think the current way this is done is now better. We still handle our own shutdown sequence in case our services won't terminate orderly, but we don't translate a SIGINT into a SIGTERM anymore and just relay the user's signal, which makes us compatible again with external process supervision systems.